FFmpegInterop/SampleProvider.cpp

```diff
@@ -0,0 +1,415 @@
+			{
+				if (masteringDisplayMetadata->has_luminance)
+				{
+					constexpr uint32_t MASTERING_DISP_LUMINANCE_SCALE{ 10000 };
+					properties.Insert(MF_MT_MIN_MASTERING_LUMINANCE, PropertyValue::CreateUInt32(static_cast<uint32_t>(MASTERING_DISP_LUMINANCE_SCALE * av_q2d(masteringDisplayMetadata->min_luminance))));
+					properties.Insert(MF_MT_MAX_MASTERING_LUMINANCE, PropertyValue::CreateUInt32(static_cast<uint32_t>(av_q2d(masteringDisplayMetadata->max_luminance))));
+				}
+
+				if (masteringDisplayMetadata->has_primaries)
+				{
+					MT_CUSTOM_VIDEO_PRIMARIES customVideoPrimaries
+					{
+						static_cast<float>(av_q2d(masteringDisplayMetadata->display_primaries[0][0])),
+						static_cast<float>(av_q2d(masteringDisplayMetadata->display_primaries[0][1])),
+						static_cast<float>(av_q2d(masteringDisplayMetadata->display_primaries[1][0])),
+						static_cast<float>(av_q2d(masteringDisplayMetadata->display_primaries[1][1])),
+						static_cast<float>(av_q2d(masteringDisplayMetadata->display_primaries[2][0])),
+						static_cast<float>(av_q2d(masteringDisplayMetadata->display_primaries[2][1])),
+						static_cast<float>(av_q2d(masteringDisplayMetadata->white_point[0])),
+						static_cast<float>(av_q2d(masteringDisplayMetadata->white_point[1]))
+					};
+					properties.Insert(MF_MT_CUSTOM_VIDEO_PRIMARIES, PropertyValue::CreateUInt8Array({ reinterpret_cast<uint8_t*>(&customVideoPrimaries), reinterpret_cast<uint8_t*>(&customVideoPrimaries) + sizeof(customVideoPrimaries) }));
+				}
+			}
+
+			break;
+		}
+
+		case AVMEDIA_TYPE_SUBTITLE:
+		{
+			ITimedMetadataEncodingProperties subtitleEncProp{ encProp.as<ITimedMetadataEncodingProperties>() };
+
+			if (setFormatUserData && codecPar->extradata != nullptr && codecPar->extradata_size > 0)
+			{
+				// Set the format user data
+				subtitleEncProp.SetFormatUserData({ codecPar->extradata, codecPar->extradata + codecPar->extradata_size });
+			}
+
+			MediaPropertySet properties{ encProp.Properties() };
+
+			properties.Insert(MF_MT_ALL_SAMPLES_INDEPENDENT, PropertyValue::CreateUInt32(true));
+			properties.Insert(MF_MT_COMPRESSED, PropertyValue::CreateUInt32(false));
+
+			break;
+		}
+
+		default:
+			WINRT_ASSERT(false);
+			THROW_HR(E_UNEXPECTED);
+		}
+	}
+
+	void SampleProvider::Select() noexcept
+	{
+		TraceLoggingWrite(g_FFmpegInteropProvider, "StreamSelected", TraceLoggingLevel(TRACE_LEVEL_VERBOSE), TraceLoggingPointer(this, "this"),
+			TraceLoggingValue(m_stream->index, "StreamId"));
+
+		WINRT_ASSERT(!m_isSelected);
+		m_isSelected = true;
+	}
+
+	void SampleProvider::Deselect() noexcept
+	{
+		TraceLoggingWrite(g_FFmpegInteropProvider, "StreamDeselected", TraceLoggingLevel(TRACE_LEVEL_VERBOSE), TraceLoggingPointer(this, "this"),
+			TraceLoggingValue(m_stream->index, "StreamId"));
+
+		WINRT_ASSERT(m_isSelected);
+		m_isSelected = false;
+		Flush();
+	}
+
+	void SampleProvider::OnSeek(_In_ int64_t seekTime) noexcept
+	{
+		m_nextSamplePts = seekTime;

```

Comment:
>m_nextSamplePts = seekTime; [](start = 2, length = 27)

The seekTime provided here is in AV_TIME_BASE units, whereas m_nextSamplePts is assumed to be in stream specific time_base units in SampleProvider::GetSample. I'm not sure how common it is for the stream specific time_base units to not be equivalent to AV_TIME_BASE units, but we should probably be more robust here and handle that case.